app/views/Settings/LocationTrackingStatus.js

```diff
@@ -0,0 +1,184 @@
+import { config } from '../../COVIDSafePathsConfig';
+import { useLocTrackingStatus } from '../../services/hooks/useLocTrackingStatus';
+import { Reason } from '../../services/LocationService';
+import { isPlatformiOS } from '../../Util';
+
+export const LocationTrackingStatus = () => {
+  const { i18n } = useTranslation();
+  const [locTrackingStatus, setLocTrackingStatus] = useLocTrackingStatus();
+  console.log(locTrackingStatus);
+
+  const { reason, canTrack } = locTrackingStatus;
+
+  const isGPS = config.tracingStrategy === 'gps';
+
+  const hasLocTrackingPermissions =
+    reason === Reason.USER_ENABLED || reason === Reason.USER_DISABLED;
+
+  const isMissingLocPermissions =
+    reason === Reason.DEVICE_LOCATION_OFF ||
+    reason === Reason.APP_NOT_AUTHORIZED;
+
+  const getCurrentRowDirection = i18n.dir() === 'rtl' ? 'row-reverse' : 'row';
+
+  const getLocStatusElement = () => {
+    console.log(`reason: ${reason}`);
+    console.log(
+      `hasLocTrackingPermissions: ${hasLocTrackingPermissions} isMissingLocPermissions: ${isMissingLocPermissions}`,
+    );
+
+    if (hasLocTrackingPermissions) {
+      return (
+        <LocationStatusSwitch
+          canTrack={canTrack}
+          setLocTrackingStatus={setLocTrackingStatus}
+        />
+      );
+    }
+
+    if (isMissingLocPermissions) {
+      return <EnableLocationButton reason={reason} />;
+    }
+  };
+
+  if (isGPS) {
+    return (
+      <>
+        <Container
+          style={css`
+            flex-direction: ${getCurrentRowDirection};
+          `}>
+          {getLocStatusElement()}
+        </Container>
+        <Divider />
+      </>
+    );
+  } else {
+    return null;
+  }
+};
+
+const EnableLocationButton = ({ reason }) => {
+  const { t } = useTranslation();
+
+  /**
+   * Conditional toggling based on the location permission actions
+   * that are enabled per platform.
+   */
+  const requestLocationPermission = async () => {
+    /**
+     * Open the settings for the app
+     */
+    if (reason === Reason.APP_NOT_AUTHORIZED) {
+      openSettings();
+    }
+
+    /**
+     * Open Device Settings.
+     * If iOS, open Settings home screen
+     * If Android, we are able to open the device location permissions
+     */
+    if (reason === Reason.DEVICE_LOCATION_OFF) {
+      if (isPlatformiOS()) {
+        const deviceSettingsUrl = 'App-prefs:';
+        Linking.openURL(deviceSettingsUrl);
+      } else {
+        DeviceSettings.location();
+      }
+    }
+  };
+
+  return (
+    <>
+      <Button
+        onPress={requestLocationPermission}
+        label={t('label.logging_request_location_permission')}
+        style={{ width: '100%' }}
+        secondary
+        small
+      />
+    </>
+  );
+};
+
+const LocationStatusSwitch = ({ canTrack, setLocTrackingStatus }) => {
+  console.log('LocationStatusSwitch!');
+  const { t, i18n } = useTranslation();
+  const [label, setLabel] = useState('');
+
+  useEffect(() => {
```

Comment:
Extract this whole text mutation/animation thing into a hook/component so that the state and timers are all isolated.

```jsx
const loadingLabel = useInProgressText(label, 1000);

// or

<Typography><AnimatedText value={label} interval={1000} /></Typography>
```

Reply:
I like the animated text approach. Going to throw it in `/components`, but let me know if you think it should just stay in this file.

Reply:
Thanks for pushing back on this - ended up being simple to implement and is 10x cleaner/easier to understand.